tests/test_seq2pat_batch.py

```diff
@@ -0,0 +1,284 @@
+# -*- coding: utf-8 -*-
+# SPDX-License-Identifier: GPL-2.0
+import collections
+import os
+import unittest
+
+from sequential.seq2pat import Seq2Pat, Attribute
+from sequential.utils import read_data, sort_pattern, list_to_counter, counter_to_list
+
+
+class TestSeq2PatBatch(unittest.TestCase):
+    TEST_DIR = os.path.dirname(os.path.abspath(__file__))
+    DATA_DIR = TEST_DIR + os.sep + "data" + os.sep
+
+    def test_list_to_counter(self):
+        patterns = [[11, 12, 1],
+                    [11, 12, 13, 1],
+                    [11, 13, 1],
+                    [12, 13, 1]]
+
+        counter = list_to_counter(patterns)
+
+        self.assertEqual(counter, collections.Counter({(11, 12): 1, (11, 12, 13): 1, (11, 13): 1, (12, 13): 1}))
+
+    def test_counter_to_list(self):
+        counter = collections.Counter({(11, 12): 1, (11, 12, 13): 2, (11, 13): 1, (12, 13): 1})
+
+        results = counter_to_list(counter, min_row_count=2)
+
+        self.assertEqual(results, [[11, 12, 13, 2]])
+
+    def test_batch_patterns_default_equal(self):
+        # Seq2Pat over 3 sequences
+        seq2pat = Seq2Pat(sequences=[["A", "C", "B", "A", "D"],
+                                     ["C", "B", "A"],
+                                     ["C", "A", "C", "D"]])
+
+        # Patterns
+        patterns = seq2pat.get_patterns(min_frequency=1.0)
+
+        seq2pat = Seq2Pat(sequences=[["A", "C", "B", "A", "D"],
+                                     ["C", "B", "A"],
+                                     ["C", "A", "C", "D"]], batch_size=2)
+
+        # Patterns
+        batch_patterns = seq2pat.get_patterns(min_frequency=1.0)
+
+        self.assertEqual(patterns, batch_patterns)
+
+    def test_batch_patterns_constraints_equal(self):
+        # Seq2Pat over 3 sequences
+        seq2pat = Seq2Pat(sequences=[["A", "C", "B", "A", "D"],
+                                     ["C", "B", "A"],
+                                     ["C", "A", "C", "D"]])
+
+        # Attribute - I: Price
+        price = Attribute(values=[[5, 5, 3, 8, 2],
+                                  [1, 3, 3],
+                                  [4, 5, 2, 1]])
+
+        # Attribute - II: Timestamp
+        timestamp = Attribute(values=[[1, 1, 2, 3, 3],
+                                      [3, 8, 9],
+                                      [2, 5, 5, 7]])
+
+        # Add Constraint
+        avg_constraint = 3 <= price.average() <= 4
+        seq2pat.add_constraint(avg_constraint)
+
+        # Patterns
+        patterns = seq2pat.get_patterns(min_frequency=0.8)
+
+        seq2pat = Seq2Pat(sequences=[["A", "C", "B", "A", "D"],
```

Comment:
While this does work, based on the comments above, you could get a different result, right?

Reply:
Yes, the idea is to make `min_frequency` to be proportionally smaller on each batch. But the occurrence of a pattern on a batch can be slightly larger or smaller than the desired threshold. If it is slightly smaller, we would still want to return it from the batch, so one way for doing this is to reduce the threshold to some degree. Then there will be an aggregation of patterns to make sure the finally returned patterns meet desired occurrence threshold.